fuel-vm/src/interpreter/blockchain.rs

```diff
@@ -636,8 +637,12 @@ where
             self.gas_cost,
             charge_len,
         )?;
-        let contract = super::contract::contract(self.storage, &contract_id)?;
-        let contract_bytes = contract.as_ref().as_ref();
+
+        let contract_buffer: Vec<u8> = load_contract_code_from_storage(
```

Comment:
After reviewing this change, I see that we basically have the same code as before and allocate vectors in all cases.

So basically, this change doesn't make a lot of sense=D

The main problem is why we can't do the optimal implementation and copy directly into the memory - contract offset(and blob offset, since we have the same problem for blobs `LDC` with `mode == 1` and `BLDD` opcode).

I think we need to extend the `StorageRead` trait if we want to fix this issue in a proper way. A new method will do the same as the `read` or change the `read` method itself to something like:

```
fn read(&self, key: &Type::Key, offset: usize, buf: &mut [u8])
        -> Result<Option<usize>, Self::Error>
```

Could you change this method to work with `offset` and update all code to not allocate vector, please?

Also, it would be nice if you do the same for blobs as well.

Reply:
Yes, the original idea was to have only one allocation. The only actual improvement in this PR as is now is fixing <MemoryStorage as StorageRead<ContractsRawCode>>::read`, which was panicking before due to slices of been different size being copied. 

Happy to extend the trait to read. I think it would be better to add a new function `read_from_offset`, and then change `read` to be a provided method with offset 0

Reply:
Sure, just to make sure: if we opt for modifying the `StorageRead::read` function, then there will be several places where the implementation needs to be updated, both in `fuel-vm` and `fuel-core`; also it is not clear to me that we can read from an arbitrary offset easily in all the implementations of the `StorageRead` trait. But I will try and see how far can I go before getting into trouble